lexibank_bds.py

```diff
@@ -0,0 +1,138 @@
+import pathlib
+from collections import defaultdict
+
+import attr
+from lingpy import Wordlist
+from collabutils.edictor import fetch
+from tabulate import tabulate
+from pylexibank import Dataset as BaseDataset, Language, Lexeme
+from clldutils.misc import slug
+
+
+@attr.s
+class Word(Lexeme):
+    autoborid = attr.ib(
+        default=None,
+        converter=lambda s: None if s in ('0', 0) else int(s),
+        metadata={'dc:description': 'Automatically inferred borrowing Identifier'}
+    )
+    autocogid = attr.ib(
+        default=None,
+        metadata={'dc:description': 'Automatically inferred cognate class'}
+    )
+    autocogids = attr.ib(
+        default=None,
+        metadata={'dc:description': 'Automatically inferred cognate classes'}
+    )
+
+
+@attr.s
+class Doculect(Language):
+    Dataset = attr.ib(default=None)
+    ChineseName = attr.ib(default=None)
+    Family = attr.ib(default=None)
+    SubGroup = attr.ib(default=None)
+    Datasets = attr.ib(default=None)
+
+
+class Dataset(BaseDataset):
+    dir = pathlib.Path(__file__).parent
+    id = "borrowing-detection-study"
+    language_class = Doculect
+    lexeme_class = Word
+
+    _wlname = 'seabor.sqlite3'
+
+    def wl(self):
+        return Wordlist(str(self.raw_dir / self._wlname))
+
+    def cldf_specs(self):  # A dataset must declare all CLDF sets it creates.
+        return super().cldf_specs()
+
+    def cmd_download(self, args):
+        fetch("seabor", outdir=self.raw_dir, remote_dbase=self._wlname, base_url="https://digling.org/edictor/")
+        wl = self.wl()
+        datasets = defaultdict(lambda: defaultdict(list))
+        langs = defaultdict(lambda: defaultdict(list))
+        for idx in wl:
+            datasets[wl[idx, 'dataset']][wl[idx, 'concept']] += [wl[idx, 'doculect']]
+            langs[wl[idx, 'dataset']][wl[idx, 'doculect']] += [wl[idx, 'concept']]
+
+        tab = []
+        for dataset in datasets:
+            tab += [[dataset, len(datasets[dataset]), len(langs[dataset])]]
+        print(tabulate(tab))
+
+    def cmd_makecldf(self, args):
+        from lingrex.cognates import common_morpheme_cognates
+        from lingrex.borrowing import internal_cognates, external_cognates
+
+        #wl = self.wl()
+        #internal_cognates(wl, runs=10000, ref="autocogids", partial=True,
+        #          method="lexstat", threshold=0.55, cluster_method="infomap")
+        #common_morpheme_cognates(wl, ref="autocogid", cognates="autocogids",
+        #                 morphemes="automorphemes")
+        #external_cognates(wl, cognates="autocogid", ref="autoborid")
```

Comment:
Oops. Needs to commented-in for final run.

Reply:
can we add a param here, that we allow to re-run or not? Or do we include this as part of the download command?